Components: Fix color source mode not being saved

rafaelgallani opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
This aims to fix #11790. Adds the preferred color picker mode as an editor setting.

## How has this been tested?
Only with manual interactions in the post pages: switch color source mode and reopening color picker; changing color source mode on the sidebar and testing the inline color source mode; etc.

## Types of changes
Bug fix.
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [X] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

rafaelgallani committed:
Try: fix color source mode not being saved

rafaelgalani commented:
@gziolo What do you think? Is this a good way of persisting this type of data? I thought of storing it as a setting to make it visible between other instances or maybe elsewhere. 
There still some work to be done, but that was the main idea I had initially. I'm still getting into React/Redux, so if that's not a good way of doing it, please let me know if there's any reference or examples I can refer to... 
Also, I didn't know which other people to tag, but any feedback/comments would be great. 😸

gziolo commented:
I think this is a good direction. You probably shouldn't be using `useEditorFeature` because it's reserved for bigger features that are related to the block controls, theme options and so on, but you can get the setting with `useSelect` and `getSettings`.

rafaelgallani committed:
Removing references to `useEditorFeature`

rafaelgallani committed:
Reverting back to `state` prop

rafaelgallani committed:
Adding source change event to components

rafaelgallani committed:
Updating tests to use new callback name

rafaelgallani commented:
It took me a while for being able to focus on this PR, sorry about it... Anyway, some changes were made: 
- `useEditorFeature()` call was replaced by `getSettings()`, as recommended by @gziolo;
- I split changes between `source` and `color` props, as these shouldn't be coupled (e.g. the source mode can be changed without implying in a color's change) - two different events were added, `onColorChange` and `onSourceChange`;
- Unit tests were added and refactored as needed - same for the test snapshots.

Is it safe to change an event's name like so? Or should this type of change be avoided? 
I thought it was indeed necessary since I don't see `source` and `color` properties as dependants.

gziolo commented on packages/components/src/color-palette/index.js:
```diff
@@ -21,10 +21,14 @@ export default function ColorPalette( {
 	disableCustomColors = false,
-	onChange,
+	colorPickerMode = 'hex',
+	onColorChange,
```
It looks like a breaking change for `onChange`, see: https://developer.wordpress.org/block-editor/how-to-guides/backward-compatibility/#how-to-preserve-backward-compatibility-for-a-react-component

Public API for this component is documented in https://github.com/WordPress/gutenberg/tree/079ef9ea0d349536cb2f5fd0a9810508bc4f1261/packages/components/src/color-palette

rafaelgallani replied:
Thanks for the guidance! 😸 
I'll try to preserve the naming/method as it's now and adding the new one.

rafaelgallani committed:
Adding new sourceChange event to gradient components

rafaelgallani committed:
Try: adding color picker mode update to color edit hooks

rafaelgallani commented:
I just managed to get some free time now by morning... 😅 I think it looks better now. Backward compatibility is preserved - it doesn't break any of the previous methods and the new ones were added in all the needed components... 
Also, I tested locally with the gradient color picker and the regular color picker, LGTM - is there another way of changing text color?

Some e2e tests would be great here too, I think I'll add a few.
